[go_router] Adds onException to GoRouter constructor. (#4216)
flutter/packages: packages/go_router/doc/error-handling.md

There are several kinds of errors or exceptions in go_router.

* GoError and AssertionError 

This kind of errors are thrown when go_router is used incorrectly, for example, if the root
[GoRoute.path](https://pub.dev/documentation/go_router/latest/go_router/GoRoute/path.html) does
not start with `/` or a builder in GoRoute is not provided. These errors should not be caught and
must be fixed in code in order to use go_router.

* GoException

This kind of exception are thrown when the configuration of go_router cannot handle incoming requests
from users or other part of the code. For example, an GoException is thrown when user enter url that
can't be parsed according to pattern specified in the `GoRouter.routes`. These exceptions can be

> handled in various callback.

Reviewer: ```suggestion
handled in various callbacks.
```

Once can provide a callback to `GoRouter.onException` to handle this exception. In this callback,

> one can choose to ignore, redirect, or push different page based on needs.

Reviewer: ```suggestion
one can choose to ignore, redirect, or push different pages depending on the situation.
```

There are several kinds of errors or exceptions in go_router.

* GoError and AssertionError 

> This kind of errors are thrown when go_router is using incorrectly, for example, if the root

Reviewer: ```suggestion
This kind of errors are thrown when go_router is used incorrectly, for example, if the root
```
